### PR TITLE
Dynamo+LTC: merging related code from staging branch to master

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75046

Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.

Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

release notes: lazy
topics: not user facing

Differential Revision: [D35300646](https://our.internmc.facebook.com/intern/diff/D35300646)

### CODE REVIEW

**Comment by shunting314** [comment]
Previous comments is in https://github.com/pytorch/pytorch/pull/75044

**Review by wconstab — APPROVED** [review]
LGTM pending some fixes commented in the previous PR.  Also pending if @Krovatkin wants to review.

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).

File: torch/csrc/lazy/python/init.cpp [hunk 1]
```diff
@@ -17,6 +20,12 @@
 namespace torch {
 namespace lazy {
 
+struct NoGilSection {
```
Comment (Krovatkin) [inline torch/csrc/lazy/python/init.cpp:23]: we should be using the pybind11 equivalent : `py::gil_scoped_release `

File: torch/_lazy/computation.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import torch._C._lazy
```
Comment (Krovatkin) [inline torch/_lazy/computation.py:1]: it seems a bit arbitrary that these are grouped together under `computation`.  For example, why didn't `sync_multi` make the cut into this file?

**Review by Krovatkin — APPROVED** [review]
a few nits

File: torch/_lazy/computation.py (hunk 2 above)
Comment (shunting314) [reply torch/_lazy/computation.py:1]: yes, I don't find a hard rule to decide which APIs goes to which module. This file specifically contains APIs needed to reused a compiled graph so far.  `sync_multi` is a more general API thus I didn't add that here. But it should also look good to put `sync_multi` here

File: torch/_lazy/debug.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import torch._C._lazy
+
+
+def render_ir_graph(tensors):
+    """Return a text dump of the LTC IR graph in dot format for the tensors.
+       The text can be processed by tools like dot to be rendered in pdf,png etc."""
+    return torch._C._lazy._get_tensors_dot(tensors)
+
+def dump_ir(tensors, ir_format):
```
Comment (shunting314) [inline torch/_lazy/debug.py:9]: @wconstab  I've made the requested changes in this pr.

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).

**Pull request closed without merging**